matron/src/hardware/screen/ssd1322.c

```diff
@@ -0,0 +1,392 @@
+#define write_command(x) \
+    (ssd1322_write_command(x, 0, 0))
+
+#ifdef SSD1322_USES_THREAD
+static void* ssd1322_thread_run(void * p){
+    (void)p;
+
+    static struct timespec ts = {
+            .tv_sec = 0,
+            .tv_nsec = 16670000, // 60Hz
+    };
+
+    while( spidev_buffer ){
+        if( display_dirty ){
+            ssd1322_refresh();
+            display_dirty = false;
+        }
+
+        // If this event happens right before ssd1322_refresh(),
+        // there is quite a bit of flashing. Possibly from being
+        // at a weird sync point with the hardware refresh.
+        event_post(event_data_new(EVENT_SCREEN_REFRESH));
+        
+	clock_nanosleep(CLOCK_MONOTONIC, 0, &ts, NULL);
+    }
+
+    return NULL;
+}
+#endif
+
+void ssd1322_init() {
+
+    if( pthread_mutex_init(&lock, NULL) != 0 ){
+        fprintf(stderr, "%s: pthread_mutex_init failed\n", __func__);
+        return;
+    }
+
+    spidev_buffer = calloc(8192, sizeof(uint8_t));
+    if( spidev_buffer == NULL ){
+        fprintf(stderr, "%s: couldn't allocate spidev_buffer\n", __func__);
+    }
+
+    spidev_fd = open_spi(SPIDEV_0_0_PATH);
+    if( spidev_fd < 0 ){
+        fprintf(stderr, "%s: couldn't open %s.\n", __func__, SPIDEV_0_0_PATH);
+        return;
+    }
+
+    gpio_0 = gpiod_chip_open_by_name(SSD1322_DC_AND_RESET_GPIO_CHIP);
+    gpio_dc = gpiod_chip_get_line(gpio_0, SSD1322_DC_GPIO_LINE);
+    gpio_reset = gpiod_chip_get_line(gpio_0, SSD1322_RESET_GPIO_LINE);
+
+    gpiod_line_request_output(gpio_dc, "D/C", 0);
+    gpiod_line_request_output(gpio_reset, "RST", 0);
+
+    // SSD1322 Reference Document (v1.2) P 16/60
+    // "Keep this pin pull HIGH during normal operation"
+    gpiod_line_set_value(gpio_reset, 1);
+
+    // All values copied from fbtft-ssd1322.c from monome/linux repo.
+    write_command(SSD1322_SET_DISPLAY_OFF);
+    write_command(SSD1322_SET_DEFAULT_LINEAR_GRAY_SCALE);
+    write_command_with_data(SSD1322_SET_OSCILLATOR_FREQUENCY, 0x91);
+    write_command_with_data(SSD1322_SET_MULTIPLEX_RATIO, NORNS_MUX_RATIO);
+    write_command_with_data(SSD1322_SET_DISPLAY_OFFSET, 0x00);
+    write_command_with_data(SSD1322_SET_DISPLAY_START_LINE, 0x00);
+    write_command_with_data(SSD1322_SET_VDD_REGULATOR, 0x01);
+    write_command_with_data(SSD1322_SET_DISPLAY_ENHANCEMENT_A, 0xA0, 0xFD);
+    write_command_with_data(SSD1322_SET_CONTRAST_CURRENT, 0x7F);
+    write_command_with_data(SSD1322_MASTER_CURRENT_CONTROL, 0x0F);
+    write_command_with_data(SSD1322_SET_PHASE_LENGTH, NORNS_PHASE_LENGTH);
+    write_command_with_data(SSD1322_SET_PRECHARGE_VOLTAGE, 0x1F);
+    write_command_with_data(SSD1322_SET_VCOMH_VOLTAGE, 0x04);
+    write_command(SSD1322_SET_DISPLAY_MODE_NORMAL);
+
+    if( platform() != PLATFORM_CM3 ){
+        write_command_with_data(SSD1322_SET_DUAL_COMM_LINE_MODE, 0x04, 0x11);
+    }
+    else{
+        write_command_with_data(SSD1322_SET_DUAL_COMM_LINE_MODE, 0x16, 0x11);
+    }
```

Comment:
> one issue i foresee (i'm attempting to roll a beta update that will work on all norns, shields, etc) is that shields have the screen rotated by 180 degrees, so in config.txt there is dtparam=rotate=180
> 
> it'd likely be trivial to do the rotation within matron, after some device detection

@tehn, I think these lines should cover that. Let me know if you test it with a shield and it doesn't work though.

Reply:
wow, fantastic! i'll test on a shield now